### PR TITLE
Feature/40 clear overview of submissions

### PR DESCRIPTION
**Pull request by todd-dembrey**

### CODE REVIEW

**Comment by todd-dembrey** [comment]
Very basic table, I'm half tempted to enable bootstrap as a sane default style.

image

**Comment by zerolab** [comment]
We discussed this. @chris-lawton will extract sane table styles. I am *very* much against bootstrap in a non-rapid prototyping context

File: requirements.txt [hunk 1]
```diff
@@ -13,6 +13,7 @@ wagtail_factories==0.3.0
 flake8
 
 social_auth_app_django==2.1.0
+django-tables2=1.17.1
```
Comment (zerolab) [inline requirements.txt]: ==

File: opentech/apply/funds/blocks.py [hunk 2]
```diff
@@ -1,9 +1,110 @@
...
+            # Pretend the error is a readonly input so that we get nice formatting
+            error_message = '<div class="error"><input readonly placeholder="{}"></div>'.format(errors[0])
+        else:
+            error_message = ''
+        form = super().render_form(*args, **kwargs)
+        form = '<br>'.join([self.description, form]) + error_message
+        return mark_safe(form)
+
+    def deconstruct(self):
+        return ('wagtail.core.blocks.StaticBlock', [], {})
```
Comment (zerolab) [inline opentech/apply/funds/blocks.py]: this is Wagtail 2.x dependency

File: opentech/apply/funds/blocks.py [hunk 3]
```diff
@@ -1,9 +1,110 @@
...
+        super().__init__(*args, **kwargs)
+
+    class Meta:
+        admin_text = 'Much be included in the form once.'
+
+    def render_form(self, *args, **kwargs):
+        errors = kwargs.pop('errors')
+        if errors:
+            # Pretend the error is a readonly input so that we get nice formatting
+            error_message = '<div class="error"><input readonly placeholder="{}"></div>'.format(errors[0])
```
Comment (zerolab) [inline opentech/apply/funds/blocks.py]: Add a reference to the related Wagtail issues/PRs

File: opentech/apply/funds/blocks.py [hunk 4]
```diff
@@ -1,9 +1,110 @@
...
+        return value
+
+
+class MustIncludeStatic(StaticBlock):
+    def __init__(self, *args, description='', **kwargs):
+        self.description = description
+        super().__init__(*args, **kwargs)
+
+    class Meta:
+        admin_text = 'Much be included in the form once.'
```
Comment (zerolab) [inline opentech/apply/funds/blocks.py]: "Must be included in the form only once" ?

File: opentech/apply/dashboard/templates/dashboard/dashboard.html [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{% extends "base.html" %}
+{% load render_table from django_tables2 %}
+
+{% block content %}
+<h1>Recieved Content</h1>
```
Comment (zerolab) [inline opentech/apply/dashboard/templates/dashboard/dashboard.html]: `s/Recieved/Received/`

File: opentech/apply/funds/blocks.py [hunk 6]
```diff
@@ -1,9 +1,129 @@
...
+from django.utils.text import mark_safe
 
-from opentech.apply.stream_forms.blocks import FormFieldsBlock
+from wagtail.wagtailcore.blocks import StaticBlock
+from opentech.apply.stream_forms.blocks import FormFieldsBlock, FormFieldBlock
 
 from opentech.apply.categories.blocks import CategoryQuestionBlock
 
 
+def find_duplciates(items):
```
Comment (zerolab) [inline opentech/apply/funds/blocks.py]: `s/duplciates/duplicates/`

File: opentech/apply/funds/blocks.py [hunk 7]
```diff
@@ -1,9 +1,129 @@
...
+                if block_name in duplicates:
+                    self.add_error_to_child(error_dict, i, 'info', 'Duplicate field')
+
+        if all_errors or error_dict:
+            error_dict['__all__'] = all_errors
+            raise ValidationError('Error', params=error_dict)
+
+        return value
+
+    def add_error_to_child(self, errors, child_number, field, message):
```
Comment (zerolab) [inline opentech/apply/funds/blocks.py]: PyCharm inspections suggest it is a static method

File: opentech/apply/funds/blocks.py [hunk 8]
```diff
@@ -1,9 +1,129 @@
...
+            # Issue discussed here: https://github.com/wagtail/wagtail/issues/4122
+            error_message = '<div class="error"><input readonly placeholder="{}"></div>'.format(errors[0])
+        else:
+            error_message = ''
+        form = super().render_form(*args, **kwargs)
+        form = '<br>'.join([self.description, form]) + error_message
+        return mark_safe(form)
+
+    def deconstruct(self):
+        return ('wagtail.wagtailcore.blocks.static_block.StaticBlock', (), {})
```
Comment (zerolab) [inline opentech/apply/funds/blocks.py]: redundant parantheses?

File: opentech/apply/dashboard/tables.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+from opentech.apply.funds.models import ApplicationSubmission
+
+
+class DashboardTable(tables.Table):
+    submit_time = tables.DateColumn(verbose_name="Submitted")
+    page = tables.Column(verbose_name="Fund")
+
+    class Meta:
+        model = ApplicationSubmission
+        fields = ('title', 'page', 'submit_time')
```
Comment (zerolab) [inline opentech/apply/dashboard/tables.py]: currently the title column outputs a dash regardless of the submission

Comment (todd-dembrey) [reply opentech/apply/dashboard/tables.py]: Fixed in 75dc7ea

File: opentech/apply/funds/blocks.py (hunk 7 above)
Comment (todd-dembrey) [reply opentech/apply/funds/blocks.py]: It is, but what value does defining it as static give us?

Comment (zerolab) [reply opentech/apply/funds/blocks.py]: A less shouty IDE. At the same time, it doesn't give us anything, so meh..

File: opentech/apply/dashboard/tables.py (hunk 9 above)
Comment (zerolab) [reply opentech/apply/dashboard/tables.py]: 👍

File: opentech/apply/funds/migrations/0005_applicationsubmission.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+# -*- coding: utf-8 -*-
```
Comment (zerolab) [inline opentech/apply/funds/migrations/0005_applicationsubmission.py]: these 2 migrations still need renumbering to 8 and 9, respectively

**Comment by zerolab** [comment]
Let's get it on dev for internal testing

**Pull request merged**